elastic/kibana
# [Reporting] Avoid blocking the event loop when generating a PDF

## Summary

The current version of the `pdfmake` library (`0.1.65`) synchronously creates a PDF. This can be a relatively CPU intensive task for PDFs that contain many images causing Kibana server-side to become unresponsive for a time to incoming requests.

## The approach

Set up for testing PDF generation performance using 0.3.0-alpha.0 of `pdfmake` which introduces asynchronous PDF generation. See https://github.com/elastic/kibana/issues/118260 for details of how to see the blocking behaviour of the current library rendering the PDF synchronously.

## How to test

1. Clone https://github.com/bpampuch/pdfmake
2. `cd` into the directory run `yarn && yarn build`
3. Run `yarn link`
4. `cd` to your kibana repo
5. Check out this PR and run `yarn kbn bootstrap` then run `yarn link pdfmake`
6. Start Kibana and see https://github.com/elastic/kibana/issues/118260 for instructions on how to test PDF generation behaviour.

## Additional notes

* This PR is used to demonstrate how a future version of the `pdfmake` library could be used to bypass the currently blocking behaviour. Further work/time will be needed before this is available and stable. The code used in this PR is considered "unstable".
